rosidl_generator_cpp/resource/msg__struct.hpp.em

```diff
@@ -81,55 +86,111 @@ struct @(spec.base_type.type)_
 {
   using Type = @(spec.base_type.type)_<ContainerAllocator>;
 
-@# constructors (with and without allocator)
-@[for (alloc_type, alloc_name) in [['', ''], ['const ContainerAllocator & ', '_alloc']]]@
-  @('explicit ' if alloc_type else '')@(spec.base_type.type)_(@(alloc_type + alloc_name))
-@# generate initializer lists
-@[  if alloc_name == '_alloc']@
-// *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
-@[  end if]@
 @{
-used_alloc_name = False
-leading_colon = ':'
-for field in spec.fields:
-    # dynamic arrays require allocator if available
-    if alloc_name and field.type.is_array and \
-            (field.type.array_size is None or field.type.is_upper_bound):
-        if leading_colon == ' ':
-            print(',')
-        print('  %s %s(%s)' % (leading_colon, field.name, alloc_name), end='')
-        used_alloc_name = True
-        leading_colon = ' '
-
-    # if default value is available set it in initializer list
-    elif field.type.is_primitive_type() and field.default_value is not None:
-        cpp_value = value_to_cpp(field.type, field.default_value)
-        if leading_colon == ' ':
-            print(',')
-        print('  %s %s(%s)' % (leading_colon, field.name, cpp_value), end='')
-        leading_colon = ' '
-if leading_colon == ' ':
-    print()
+# The creation of the constructors for messages is a bit complicated.  The goal
+# is to have a constructor where the user can control how the fields of the
+# message get initialized via the _init parameter to the constructor.  See
+# http://design.ros2.org/articles/generated_interfaces_cpp.html#constructors
+# for a detailed explanation of the different _init parameters.
+init_list, alloc_list, member_list = create_init_alloc_and_member_lists(spec)
+
+def generate_default_string(membset):
+    strlist = []
+    for memb in membset.members:
+        if memb.default_value is not None:
+            if memb.num_prealloc > 0:
+                strlist.append("this->%s.resize(%d);" % (memb.name, memb.num_prealloc))
+            if isinstance(memb.default_value, list):
+                if not memb.default_value or all(v == memb.default_value[0] for v in memb.default_value):
+                    strlist.append("std::fill(this->%s.begin(), this->%s.end(), %s);" % (memb.name, memb.name, memb.default_value[0]))
```

Comment:
In the case the first condition is true (`not memb.default_value`) the list is empty. Would accessing `memb.default_value[0]` not fail?

Reply:
Yeah, this is a great catch.  I'm not sure what I was trying to do now.  I've now hoisted the `if memb.default_value` up to a higher level, which should get rid of this concern.